### ISSUE TITLE
Hanging when pacman dependencies are missing

### ISSUE THREAD

**Issue by zoresvit** [rmarquis/pacaur#542]
When using `pacaur` to install an AUR package with missing dependencies from official repositories, `pacaur` just hangs on _"resolving dependencies..."._ See example with `xkb-switch-git` AUR package below:

```
zoresvit@xps:~$ pacaur -y xkb-switch-git
:: resolving dependencies...
^C
zoresvit@xps:1:~$ cower -d xkb-switch-git
:: xkb-switch-git downloaded to /home/zoresvit
zoresvit@xps:~$ cd xkb-switch-git/
zoresvit@xps:xkb-switch-git$ makepkg
==> Making package: xkb-switch-git 20150826-4 (Mon Aug 22 19:06:18 PDT 2016)
==> Checking runtime dependencies...
==> Checking buildtime dependencies...
==> Missing dependencies:
  -> cmake
==> ERROR: Could not resolve all dependencies.
```

Once the dependencies are installed manually, `pacaur` succeeds to install the AUR package.

**Comment by rmarquis** [comment]
Please follow the "Bug reports" link and provide a debug output.

**Comment by zoresvit** [comment]
@rmarquis Ooops, I'm very sorry I missed the Bug reports links. First time here.

Now that I tried from home network, `pacaur` actually works correctly and resolves dependencies.

As the first time I was trying to perform install from work, I guess the above behavior could be caused by inability to access some remote server due to network restrictions. Though download from AUR and manual installation via `pacman` was successful so it's hard to tell for sure.

Tomorrow I will try to run `pacaur` with debug output on the same network to see what stalls the process. Maybe it will help us to make `pacaur` more verbose in case of network issues.

But feel free to close this issue if you think it's not worth it, as clearly there was something wrong with the environment.

Thank you!

**Comment by rmarquis** [comment]
Thanks, I'll wait your input tomorrow. Pacaur should handle missing dependencies very well, but there might be a bug that occurs only in a very specific situation, which is the reason I need a debug output.

**Comment by zoresvit** [comment]
So I tried again on the same network and the issue was reproducible, however not related to `pacaur` (as I expected).

TL;DR
---------

`pactree` resolves dependencies using remote sync databases, calls GPG to get some key, which hangs forever on `SYN_SENT` without getting `ACK` in reply from the remote server (probably due to peculiar restrictions on that network, didn't dig deeper).

I'm surprised neither GPG nor `pactree` have some kind of timeout in case of unresponsive network. But I don't think that's something that needs to be added on such a high level as `pacaur` is either.

Full version
---------------

`pacaur` hangs on executing command `pactree -su cmake`:

```
zoresvit@xps:~$ PS4='Line ${LINENO}: ' bash -x pacaur -y xkb-switch-git 
Line 7: version=4.6.8
Line 14: unset aur cleancache devel edit info installpkg foreign needed noconfirm nodeps noedit
Line 15: unset operation pac pacQ pacS quiet rebuild refresh repo search selective upgrade
Line 18: LC_COLLATE=C
Line 19: TEXTDOMAIN=pacaur
Line 20: TEXTDOMAINDIR=/usr/share/locale
Line 23: [[ -n '' ]]
Line 28: configdir=/etc/xdg/pacaur
Line 29: userconfigdir=/home/zoresvit/.config/pacaur
Line 30: userpacmandir=/home/zoresvit/.config/pacman
Line 31: usercachedir=/home/zoresvit/.cache/pacaur
Line 34: [[ -r '' ]]
Line 37: source /etc/makepkg.conf
LLine 15: DLAGENTS=('ftp::/usr/bin/curl -fC - --ftp-pasv --retry 3 --retry-delay 3 -o %o %u' 'http::/usr/bin/curl -fLC - --retry 3 --retry-delay 3 -o %o %u' 'https::/usr/bin/curl -fLC - --retry 3 --retry-delay 3 -o %o %u' 'rsync::/usr/bin/rsync --no-motd -z %u %o' 'scp::/usr/bin/scp -C %u %o')
LLine 27: VCSCLIENTS=('bzr::bzr' 'git::git' 'hg::mercurial' 'svn::subversion')
LLine 33: CARCH=x86_64
LLine 34: CHOST=x86_64-pc-linux-gnu
LLine 39: CPPFLAGS=-D_FORTIFY_SOURCE=2
LLine 40: CFLAGS='-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong'
LLine 41: CXXFLAGS='-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong'
LLine 42: LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro
LLine 46: DEBUG_CFLAGS='-g -fvar-tracking-assignments'
LLine 47: DEBUG_CXXFLAGS='-g -fvar-tracking-assignments'
LLine 62: BUILDENV=(!distcc color !ccache check !sign)
LLine 90: OPTIONS=(strip docs !libtool !staticlibs emptydirs zipman purge !optipng !upx !debug)
LLine 93: INTEGRITY_CHECK=(md5)
LLine 95: STRIP_BINARIES=--strip-all
LLine 97: STRIP_SHARED=--strip-unneeded
LLine 99: STRIP_STATIC=--strip-debug
LLine 101: MAN_DIRS=({usr{,/local}{,/share},opt/*}/{man,info})
LLine 103: DOC_DIRS=(usr/{,local/}{,share/}{doc,gtk-doc} opt/*/{doc,gtk-doc})
LLine 105: PURGE_TARGETS=(usr/{,share}/info/dir .packlist *.pod)
LLine 130: COMPRESSGZ=(gzip -c -f -n)
LLine 131: COMPRESSBZ2=(bzip2 -c -f)
LLine 132: COMPRESSXZ=(xz -c -z -)
LLine 133: COMPRESSLRZ=(lrzip -q)
LLine 134: COMPRESSLZO=(lzop -q)
LLine 135: COMPRESSZ=(compress -c -f)
LLine 144: PKGEXT=.pkg.tar.xz
LLine 145: SRCEXT=.src.tar.gz
...
...
LLine 202: GetJson var '{"version":5,"type":"multiinfo","resultcount":1,"results":[{"ID":293669,"Name":"xkb-switch-git","PackageBaseID":98558,"PackageBase":"xkb-switch-git","Version":"20150826-4","Description":"Program that allows to query and change the XKB layout state","URL":"https:\/\/github.com\/ierton\/xkb-switch","NumVotes":13,"Popularity":1.314532,"OutOfDate":null,"Maintainer":"DeaDDooMER","FirstSubmitted":1440598225,"LastModified":1460580386,"URLPath":"\/cgit\/aur.git\/snapshot\/xkb-switch-git.tar.gz","Depends":["libx11"],"MakeDepends":["git","cmake"],"License":["GPL3"],"Keywords":["xkb-switch"]}]}' Version
LLine 1273: json_verify -q
LLine 1274: case "$1" in
LLine 1276: json_reformat
LLine 1276: tr -d '", '
LLine 1276: awk -F '#' '{print $2}'
LLine 1276: grep -Po 'Version:.*'
LLine 1276: sed -r s/Version:/Version#/g
Line 203: depsAood=($(GetJson "var" "$json" "OutOfDate"))
LLine 203: GetJson var '{"version":5,"type":"multiinfo","resultcount":1,"results":[{"ID":293669,"Name":"xkb-switch-git","PackageBaseID":98558,"PackageBase":"xkb-switch-git","Version":"20150826-4","Description":"Program that allows to query and change the XKB layout state","URL":"https:\/\/github.com\/ierton\/xkb-switch","NumVotes":13,"Popularity":1.314532,"OutOfDate":null,"Maintainer":"DeaDDooMER","FirstSubmitted":1440598225,"LastModified":1460580386,"URLPath":"\/cgit\/aur.git\/snapshot\/xkb-switch-git.tar.gz","Depends":["libx11"],"MakeDepends":["git","cmake"],"License":["GPL3"],"Keywords":["xkb-switch"]}]}' OutOfDate
LLine 1273: json_verify -q
LLine 1274: case "$1" in
LLine 1276: json_reformat
LLine 1276: tr -d '", '
LLine 1276: grep -Po 'OutOfDate:.*'
LLine 1276: sed -r s/OutOfDate:/OutOfDate#/g
LLine 1276: awk -F '#' '{print $2}'
Line 204: for i in '"${!depsAname[@]}"'
LLine 205: expac -Qs %v '^xkb-switch-git$'
Line 205: depsQver[$i]=
Line 206: [[ -z '' ]]
Line 206: depsQver[$i]='#'
LLine 207: grep -E '\-(cvs|svn|git|hg|bzr|darcs|nightly.*)$'
Line 207: [[ -n xkb-switch-git ]]
Line 207: depsAver[$i]=latest
Line 214: [[ -n '' ]]
Line 217: FindDepsRepo cmake
Line 396: local allrepopkgs providersrepopkgs providersrepopkgsrm i j
Line 398: [[ -z cmake ]]
Line 401: repodeps=($(tr ' ' '\n' <<< ${repodeps[@]} | sort -u))
LLine 401: tr ' ' '\n'
LLine 401: sort -u
Line 403: for i in '"${repodeps[@]}"'
Line 404: allrepopkgs+=($(pactree -su "$i"))
LLine 404: pactree -su cmake
```

By running `strace` I saw that it timeouts on `select` call, likely related to GPG:

```
zoresvit@xps:~$ strace pactree -su cmake
...
...
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f0351ed3a10) = 9549
wait4(9549, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 9549
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=9549, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
close(5)                                = 0
read(4, "gpg (GnuPG) 2.1.15\nlibgcrypt 1.7"..., 79) = 79
close(4)                                = 0
pipe([4, 5])                            = 0
fcntl(4, F_SETFD, FD_CLOEXEC)           = 0
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(1, TCGETS, {B38400 opost isig icanon echo ...}) = 0
fstat(1, {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 5), ...}) = 0
readlink("/proc/self/fd/1", "/dev/pts/5", 63) = 10
stat("/dev/pts/5", {st_mode=S_IFCHR|0620, st_rdev=makedev(136, 5), ...}) = 0
pipe([6, 7])                            = 0
fcntl(6, F_SETFD, FD_CLOEXEC)           = 0
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f0351ed3a10) = 9551
wait4(9551, [{WIFEXITED(s) && WEXITSTATUS(s) == 0}], 0, NULL) = 9551
--- SIGCHLD {si_signo=SIGCHLD, si_code=CLD_EXITED, si_pid=9551, si_uid=1000, si_status=0, si_utime=0, si_stime=0} ---
close(5)                                = 0
close(7)                                = 0
select(7, [4 6], [], NULL, {1, 0})      = 0 (Timeout)
select(7, [4 6], [], NULL, {1, 0})      = 0 (Timeout)
select(7, [4 6], [], NULL, {1, 0})      = 0 (Timeout)
```

Checked opened file descriptors for the process:

```
zoresvit@xps:~$ pgrep pactree
9576
zoresvit@xps:~$ ls -al /proc/9576/fd
total 0
dr-x------ 2 zoresvit zoresvit  0 Aug 23 15:02 .
dr-xr-xr-x 9 zoresvit zoresvit  0 Aug 23 15:02 ..
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:03 0 -> /dev/pts/5
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:02 1 -> /dev/pts/5
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:03 2 -> /dev/pts/5
lr-x------ 1 zoresvit zoresvit 64 Aug 23 15:03 3 -> /etc/pacman.conf
lr-x------ 1 zoresvit zoresvit 64 Aug 23 15:03 4 -> 'pipe:[672164]'
lr-x------ 1 zoresvit zoresvit 64 Aug 23 15:03 6 -> 'pipe:[673065]'
```

Tried to figure out what the opened pipes relate to, turns out it's indeed GPG:
```
zoresvit@xps:~$ sudo lsof | grep 672164
pactree    9576               zoresvit    4r     FIFO               0,10       0t0     672164 pipe
gpg2       9882               zoresvit    5w     FIFO               0,10       0t0     672164 pipe

zoresvit@xps:~$ ps aux | grep 9882
zoresvit  9882  0.0  0.0  35136  1952 pts/5    SL+  15:10   0:00 gpg2 --batch --no-sk-comments --lc-messages en_US.UTF-8 --lc-ctype en_US.UTF-8 --status-fd 5 --no-tty --charset utf8 --enable-progress-filter --display :0 --ttyname /dev/pts/5 --ttytype rxvt-unicode-256color --with-colons --fixed-list-mode --with-fingerprint --with-fingerprint --search-keys -- 0x962DDE58
```

Looked at file descriptors opened by GPG in order to figure out what's blocking it:

```
zoresvit@xps:~$ ls -al /proc/9882/fd
total 0
dr-x------ 2 zoresvit zoresvit  0 Aug 23 15:11 .
dr-xr-xr-x 9 zoresvit zoresvit  0 Aug 23 15:10 ..
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:11 0 -> /dev/null
l-wx------ 1 zoresvit zoresvit 64 Aug 23 15:11 1 -> 'pipe:[673065]'
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:11 2 -> /dev/null
lrwx------ 1 zoresvit zoresvit 64 Aug 23 15:11 3 -> 'socket:[675031]'
l-wx------ 1 zoresvit zoresvit 64 Aug 23 15:11 5 -> 'pipe:[672164]'
```

The opened socket turned out to be local one between GPG and `dirmngr`: 
```
zoresvit@xps:~$ sudo ss -axnp | grep 689825
u_str  ESTAB      0      0         * 689825                * 691721              users:(("gpg2",pid=10373,fd=3))
u_str  ESTAB      0      0      /run/user/1000/gnupg/S.dirmngr 691721                * 689825              users:(("dirmngr",pid=6776,fd=0)
```

And finally looked what's `dirmngr` is trying to do and realized it's not getting `ACK` from the server:

```
zoresvit@xps:~$ sudo lsof -i | grep 6776
dirmngr   6776 zoresvit    5u  IPv4 693303      0t0  TCP xps.**************:39592->host-37-191-220-247.lynet.no:hkp (SYN_SENT)
```

**Comment by rmarquis** [comment]
Interesting, I wasn't aware that pactree relayed on PGP. Still not sure why it does that, since that shouldn't be required to revtrieve dependencies... This might be a pactree bug, you might want to report it upstream.
